File: ode/cr_3-body_problem/julia/3-body_orbit.jl
# Restricted 3-body problem in corotating frame
using DifferentialEquations
using Plots

include("config.jl")

r_1 = u -> hypot(u[1] + mu, u[2], u[3])
r_2 = u -> hypot(u[1] - 1 + mu, u[2], u[3])

# Numerical Methods for Ordinary Differential Equations
# J. C. Butcher
# ISBN: 9781119121503
function cr3body(du, u, p, t)
    du[1] = u[4]
    du[2] = u[5]
    du[3] = u[6]
    du[4] = 2*u[5] + u[1] - (1 - p)*(u[1] + p)/r_1(u)^3 - p*(u[1] + p - 1)/r_2(u)^3
    du[5] = -2*u[4] + u[2] - (1 - p)*u[2]/r_1(u)^3 - p*u[2]/r_2(u)^3
    du[6] =  -(1 - p)*u[3]/r_1(u)^3 - p*u[3]/r_2(u)^3
end

# Jacobi constant
jacobi(u, p) = 0.5*(u[4]^2 + u[5]^2 + u[6]^2) - (1 - p)/r_1(u) - p/r_2(u) - 0.5*((1 - p)*r_1(u)^2 + p*r_2(u)^2)

t_step = t_period/N
prob = ODEProblem(cr3body, u0, (0.0, t_period), mu)
sol = solve(prob, Vern7(), adaptive=false, dt=t_step)

println("mu = ", mu)
println("period = ", t_period)
# display(sol)
data = reduce(hcat, sol.u)'

gr(size = (300, 300))
Plots.scalefontsizes()
Plots.scalefontsizes(0.4)
plotOrbit = scatter(
    data[:,1], data[:,2],
    aspect_ratio = 1,
    xlabel = "x",
    ylabel = "y",
    plot_title = "Restricted 3-body problem",
    markersize = 0.25,
    alpha = 0.05,
    legend = false,
    dpi = 300,
    annotation = (0.15, -0.25, text("\$\\mu= $(mu)\$\n\$T = 0...$(t_period)\$", :crimson, :bottom, 4))
)

# Heavy body position
scatter!(
    plotOrbit,
    [-mu], [0],
    markersize = 5.0,
    markercolor = :steelblue,
    alpha = 0.8
)

# Light body position
scatter!(
    plotOrbit,
    [1 - mu], [0],
    markersize = 2.0,
    markercolor = :orange,
    alpha = 0.8
)

plot!(plotOrbit, x_foreground_color_axis=:lightgrey, y_foreground_color_axis=:lightgrey)
plot!(plotOrbit, x_foreground_color_text=:lightgrey, y_foreground_color_text=:lightgrey)
plot!(plotOrbit, x_foreground_color_border=:lightgrey, y_foreground_color_border=:lightgrey)

#gui(), readline()
savefig(plotOrbit, "orbit.png")

inv0 = jacobi(u0, mu)
println("jacobi constant = ", inv0)
inv = map(jacobi, sol.u, map(x -> mu, sol.t))

plotError = scatter(
    sol.t, (inv .- inv0)./inv0,
    xlabel = "time",
    ylabel = "error",
    plot_title = "Relative invariant error in restricted 3-body problem",
    markersize = 0.8,
    alpha = 0.15,
    legend = false,
    dpi = 300,
    annotation = (t_period - 3.0, 0.0, text("\$\\mu= $(mu)\$\n\$inv = $(inv0)\$", :crimson, :bottom, 4))
)

# gui(), readline()
savefig(plotError, "error.png")

File: ode/cr_3-body_problem/python/3-body_orbit.py
# Circular restricted 3-body problem in corotating frame
import numpy as np
import matplotlib.pyplot as plt

from numpy import sqrt
from scipy.integrate import solve_ivp
from matplotlib.patches import Circle

import config as cfg

# Distance to the big body
def r_1(x):
    x_1, x_2, x_3 = x
    mu = cfg.mu
    return sqrt((x_1 + mu)**2 + x_2**2 + x_3**2)

# Distance to the small body
def r_2(x):
    x_1, x_2, x_3 = x
    mu = cfg.mu
    return sqrt((x_1 + mu - 1)**2 + x_2**2 + x_3**2)

# The Jacobi constant
def energy(z):
    x_1, x_2, x_3, u_1, u_2, u_3 = z
    x = [x_1, x_2, x_3]
    mu = cfg.mu
    return 0.5*(u_1**2 + u_2**2 + u_3**2) - (1 - mu)/r_1(x) - mu/r_2(x) - 0.5*((1 - mu)*r_1(x)**2 + mu*r_2(x)**2)

# Differential equations
def dfunc(t, z):
    x_1, x_2, x_3, u_1, u_2, u_3 = z
    x = [x_1, x_2, x_3]
    mu = cfg.mu
    dx_1 = u_1
    dx_2 = u_2
    dx_3 = u_3
    du_1 = 2*u_2 + x_1 - (1 - mu)*(x_1 + mu)/r_1(x)**3 - mu*(x_1 - 1 + mu)/r_2(x)**3
    du_2 = -2*u_1 + x_2 - (1 - mu)*x_2/r_1(x)**3 - mu*x_2/r_2(x)**3
    du_3 = -(1 - mu)*x_3/r_1(x)**3 - mu*x_3/r_2(x)**3
    return [dx_1, dx_2, dx_3, du_1, du_2, du_3]

# Solve equations of motion
e0 = energy(cfg.z_init)
print("J = {}".format(e0))
sol = solve_ivp(dfunc, [0, cfg.t_max], cfg.z_init, t_eval=np.arange(0, cfg.t_max, cfg.dt), method="DOP853", rtol=1e-10, atol=1e-10)

# Plot orbit in corotating frame
fig = plt.figure(figsize=(8, 8))
plt.title("Circular resticted 3-body problem")
plt.axis("equal")
plt.xlabel("x")
plt.ylabel("y")
plt.grid(True)

plt.plot(sol.y[0], sol.y[1], fillstyle='none', marker='o', markersize=1.0, linestyle='None', alpha=0.25)
body1 = Circle((-cfg.mu, 0), radius=0.05, color='darkslateblue', alpha=0.5)
body2 = Circle((1 - cfg.mu, 0), radius=0.02, color='crimson', alpha=0.5)
plt.gca().add_artist(body1)
plt.gca().add_artist(body2)
xmin, xmax, ymin, ymax = plt.axis()
plt.gca().set_xlim([xmin, max(xmax, 1.05)])
plt.text(xmin + 0.01*(xmax - xmin), ymax - 0.05*(xmax - xmin), "mu = {:.6f}\nt_max = {:.6f}".format(cfg.mu, cfg.t_max))
fig.savefig('orbit.png', dpi=150)

# Plot relative error in Jacobi constant value
fig = plt.figure(figsize=(8, 8))
plt.title("CR3BP. Invariant error")
plt.xlabel("Time")
plt.ylabel("Error")
plt.grid(True)

plt.plot(sol.t, energy(sol.y) - e0)
plt.legend(['Energy'])
fig.savefig('error.png', dpi=150)
